piker/clearing/_ems.py

```diff
@@ -638,66 +674,181 @@ async def _emsd_main(
+      brokderd trading api.
+       |
+        - ``clear_dark_triggers()``:
+          run (dark order) conditions on inputs and trigger brokerd "live"
+          order submissions.
+       |
+        - ``translate_and_relay_brokerd_events()``:
+          accept normalized trades responses from brokerd, process and
+          relay to ems client(s); this is a effectively a "trade event
+          reponse" proxy-broker.
+       |
+        - ``process_client_order_cmds()``:
+          accepts order cmds from requesting piker clients, registers
+          execs with exec loop
 
-    - ``exec_loop()``:
-      run (dark) conditions on inputs and trigger broker submissions
+    """
+    # from ._client import send_order_cmds
 
-    - ``process_broker_trades()``:
-      accept normalized trades responses, process and relay to ems client(s)
+    global _router
+    dark_book = _router.get_dark_book(broker)
 
-    """
-    from ._client import send_order_cmds
+    ems_ctx = ctx
 
-    dark_book = get_dark_book(broker)
+    cached_feed = _router.feeds.get((broker, symbol))
+    if cached_feed:
+        # TODO: use cached feeds per calling-actor
+        log.warning(f'Opening duplicate feed for {(broker, symbol)}')
 
     # spawn one task per broker feed
-    async with trio.open_nursery() as n:
+    async with (
+        trio.open_nursery() as n,
 
         # TODO: eventually support N-brokers
-        async with data.open_feed(
+        data.open_feed(
             broker,
             [symbol],
-            loglevel='info',
-        ) as feed:
-
-            # get a portal back to the client
-            async with tractor.wait_for_actor(client_actor_name) as portal:
-
-                # connect back to the calling actor (the one that is
-                # acting as an EMS client and will submit orders) to
-                # receive requests pushed over a tractor stream
-                # using (for now) an async generator.
-                async with portal.open_stream_from(
-                    send_order_cmds,
-                    symbol_key=symbol,
-                ) as order_stream:
-
-                    # start the condition scan loop
-                    quote, feed, client = await n.start(
-                        exec_loop,
-                        ctx,
-                        feed,
-                        broker,
-                        symbol,
-                        _mode,
-                    )
+            loglevel=loglevel,
+        ) as feed,
+    ):
+        if not cached_feed:
+            _router.feeds[(broker, symbol)] = feed
+
+        # XXX: this should be initial price quote from target provider
+        first_quote = await feed.receive()
+
+        # open a stream with the brokerd backend for order
+        # flow dialogue
+
+        book = _router.get_dark_book(broker)
+        book.lasts[(broker, symbol)] = first_quote[symbol]['last']
+
+        trades_endpoint = getattr(feed.mod, 'trades_dialogue', None)
```

Comment:
eventually we'll want to support independent feeds from data vs. broker providers.
this will probably be something to dig into as we get aggregate feeds going 😎